### PR TITLE
Feature/better gates - Display name and gate spawning for E2 & SF

### PR DESCRIPTION
**Pull request by snepderg**
Adds 'two' (have to have multiple for optional arguments) functions to e2helper.
- getGateName: Gets the "action" name of a Wire Gate entity.
- gateSpawn: Allows users to spawn Wire Gates given a specified type.

### CODE REVIEW

**Comment by banoargeek** [comment]
Are **nocollide** and **frozen** booleans, or are they numbers? I'm having trouble figuring this one out.

**Comment by banoargeek** [comment]
There's still a lot of uncertainty here, I have a feeling trying to call these functions will cause errors. I don't want to test today.

**Comment by banoargeek** [comment]
I still need to update the helper O_o

**Comment by banoargeek** [comment]
hiatus.

**Comment by banoargeek** [comment]
Yeah... it doesn't work but it doesn't error. Phatso said I should do some test prints, but I think the problem lies in existing WireLib funcs.

File: lua/entities/gmod_wire_expression2/core/custom/cfc_gates.lua [hunk 1]
```diff
@@ -0,0 +1,35 @@
+CFCE2Lib.RegisterExtension( "cfc_e2_lib", true )
+
+local DEFAULT_GATE_MODEL = "models/jaanus/wiretool/wiretool_gate.mdl"
+
+e2function string entity:getGateName()
+    if not IsValid( this ) or not baseclass.Get( this ) then return "" end
```
[the remark this answers is not in the document]
Comment (plally) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_gates.lua:6]: you might even be able to get rid of the baseclass check and do `return this.WireDebugName or ""`

**Review by plally — DISMISSED** [review]

**Comment by brandonsturgeon** [comment]
@banoargeek Interested in finishing this up?

**Comment by banoargeek** [comment]
Oof, I didn't realize how long this has been open for. I'm looking into it now.

File: lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua [hunk 2]
```diff
@@ -60,4 +60,5 @@ e2function void entity:ejectPodTo( vector pos )
 
     driver:ExitVehicle()
     driver:SetPos( clampedPos )
-end
\ No newline at end of file
+end
+
```
[the remark this answers is not in the document]
Comment (banoargeek) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua:64]: I don't believe I touched the entity file, but I removed the extra newline and added one to the gates file.

File: lua/entities/gmod_wire_expression2/core/custom/cfc_gates.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (banoargeek) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_gates.lua]: I'm unsure about checking if the baseclass is something specific, but I changed it to the suggestion you made. I still have to test the gate spawning and figure out what I didn't last year so I'll be sure to test the whole thing while I'm at it

File: lua/entities/gmod_wire_expression2/core/custom/cl_cfc_e2_lib.lua [hunk 3]
```diff
@@ -69,3 +69,6 @@ E2Helper.Descriptions["npcGetGlobalSquad(e:)"]         = "Returns an NPC's squad
 E2Helper.Descriptions["npcGetWeaponProficiency(e:)"]   = "Returns the weapon proficiency of the NPC."
 E2Helper.Descriptions["npcGetDamageMultiplier(e:)"]    = "Returns the damage multiplier of the NPC"
 E2Helper.Descriptions["npcCouldNotReach(e:e)"]         = "Returns 1 if the NPC marked ent2 as unreachable"
+
+-- Wire Gate Functions
+E2Helper.Description["getGateName(e:)"]             = "Returns the display name of a Wire Gate."
```
[the remark this answers is not in the document]
Comment (banoargeek) [reply lua/entities/gmod_wire_expression2/core/custom/cl_cfc_e2_lib.lua]: Good call

**Comment by banoargeek** [comment]
gateSpawn needs to be tested, I don't know why it didn't work before because I couldn't find an internal method for wire gates as SENTs

**Comment by banoargeek** [comment]
Sorry for the delay, still not sure how to tackle this problem unfortunately, and I've been busy and/or unmotivated

**Comment by banoargeek** [comment]
(There are also conflicts to work out)

**Comment by banoargeek** [comment]
School's been getting in the way, how soon would you like this done? I just need to find a good way to test the spawning functions and we should be good. Right now I don't have the knowledge to figure out what's going wrong.

**Comment by banoargeek** [comment]
(Meanwhile I'm working on figuring out the name functions, that should be the easier part I reckon)

**Comment by banoargeek** [comment]
I don't like the solution for getting a gate name. It doesn't imply that it returns a table containing both name and group name. Sure, I can update the e2 helper but I still think it's less user intuitive to do it this way.

**Comment by banoargeek** [comment]
Huh. I'm not seeing a good way to fetch the gate's action name since it's in another table and there's no reference to it within the Action table of a gate entity.

**Comment by banoargeek** [comment]
Update:
- Fixed getting action name
- Fixed spawning funcs (sorta)
 * Currently I haven't been able to figure out how to pass nocollide and frozen to `WireLib.MakeWireGate` properly, and it has some weird parameter called `noclip` as well.

**Pull request closed without merging**